File: LifeCycle/lifecycle/java/src/cdmw/lifecycle/EntityCleanupNotifier.java
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES 
 * and SELEX-SI. 
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003. 
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under 
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your 
 * option) any later version. 
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public 
 * License for more details. 
 * 
 * You should have received a copy of the GNU Library General 
 * Public License along with CARDAMOM; see the file COPYING. If not, write to 
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


package cdmw.lifecycle;

import cdmw.ossupport.ReadWriteLock;
import cdmw.common.Assert;

import com.thalesgroup.CdmwLifeCycle.EntityStrategy.EntityDatum;
import com.thalesgroup.CdmwLifeCycle.EntityStrategy.EntityDataHelper;

import com.thalesgroup.CdmwNamingAndRepository.NameDomain;
import com.thalesgroup.CdmwNamingAndRepository.InvalidName;
import com.thalesgroup.CdmwNamingAndRepository.NameDomainPackage.NotRegistered;

import org.omg.CosPropertyService.PropertySet;
import org.omg.CosPropertyService.InvalidPropertyName;
import org.omg.CosPropertyService.PropertyNotFound;

/**
 * This class provides a Cleanup Notifier for Life Cycle Framework
 * with "Entity" strategy. It maintains (object key, object Id) associations
 * of CDMW objects registered with a NameDomain. This enables unregistration
 * of the object key from the NameDomain when needed.</p>
 *
 * TODO: Add a ReadersWriter lock for Thread Safety
 *
 */
public class EntityCleanupNotifier extends CleanupNotifier {

    /**
     * Object reference of the NameDomain where CDMW objects whose
     * (oid,key) pairs are stored in self are registered.
     */
    private NameDomain domain;

    /**
     * Internal Entries Database.
     */
    private EntriesMap entries = new EntriesMap();

    /**
     * Property set for state storage/retreival
     */
    private PropertySet propertySet = null;

    /**
     * Name of the property within the property set
     */
    private String propertyName = new String();

    /**
     * Readers/Writer Lock to protect from concurrent access to internal data
     */
    private ReadWriteLock readWriteLock = new ReadWriteLock();

    /**
     * This class is used to maintain (object key, object Id) associations
     */
    class EntriesMap {

        private java.util.HashMap map = new java.util.HashMap();

        public byte[] getObjectId(String key) {
            return ( (String) map.get(key) ).getBytes();
        }

        public String getKey(byte[] oid) {
            java.util.Set entries = map.entrySet();
            String objectId = new String(oid);
            for (java.util.Iterator i = entries.iterator() ; i.hasNext() ; ) {
                java.util.Map.Entry entry = (java.util.Map.Entry) i.next();
                if ( entry.getValue().equals(objectId) ) {
                    return (String) entry.getKey();
                }
            }
            return null;
        }

        public void insertEntry(byte[] oid, String key) {
            // NB: the first value of the HashMap is the key
            map.put( key, new String(oid) );
        }

        public void removeByObjectId(byte[] oid) {
            map.remove(getKey(oid));
        }

        public void removeByKey(String key) {
            map.remove(key);
        }

        public boolean isEmpty() {
            return map.isEmpty();
        }

        public int size() {
            return map.size();
        }

        public boolean containsObjectId(byte[] oid) {
            return this.map.containsValue( new String(oid) );
        }

        public boolean containsKey(String key) {
            return this.map.containsKey( key );
        }

        public String dump() {
            java.util.Set entries = map.entrySet();
            StringBuffer output = new StringBuffer();
            for (java.util.Iterator i = entries.iterator() ; i.hasNext() ; ) {
                java.util.Map.Entry entry = (java.util.Map.Entry) i.next();
                output.append("<ObjectId = " + (String) entry.getValue() +
                    ", Key = " + (String) entry.getKey() + ">\n");
            }
            return output.toString();
        }

        public EntityDatum[] toEntityData() {
            java.util.Set entries = map.entrySet();
            EntityDatum[] ed = new EntityDatum[map.size()];
            int j = 0;
            for (java.util.Iterator i = entries.iterator() ; i.hasNext() ; ) {
                java.util.Map.Entry entry = (java.util.Map.Entry) i.next();
                byte[] oid = ((String) entry.getValue()).getBytes();
                String key = (String) entry.getKey();
                ed[j] = new EntityDatum( oid, key );
                j++;
            }
            return ed;
        }

    }


    /**
     * Constructor
     *
     * @param domain The CDMW NameDomain where stored object keys are
     * registered.
     */
    public EntityCleanupNotifier(NameDomain domain) {
        this.domain = domain;
    }

    /**
     * Unregisters the objet key associated to the specified <I>oid</I> from
     * the NameDomain and deletes its entry from the internal entries database.
     *
     * @param oid The Object Id of the CDMW object registered
     * with the NameDomain.
     */
    public void unregisterObject(byte[] oid) {

        Assert.check( entries!=null );
        readWriteLock.writeLock();
        
        boolean error = false;
        try {
            String key = entries.getKey(oid);
            if (key != null) {                
                this.domain.release_name(key);
                this.entries.removeByObjectId(oid);
            } else {
                error = true;
            }
        } catch (NotRegistered nr) {
            error = true;
            // This should not happen (TODO): ADD Trace logs
        } catch (InvalidName in) {
            error = true;
            // This should not happen either (TODO): ADD Trace logs
        } finally {
            readWriteLock.writeUnlock();
        }
        
        if (error) {
            System.out.println("Dumping table...");
            dump(System.out);
        }

    }

    /**
     * Inserts an (oid,key) pair to the internal database.
     *
     * @param oid The Object Id of a CDMW object registered
     * within the NameDomain.
     * @param key The Object Key used for the object's registration
     * in the NameDomain.
     *
     * @return <B>true</B> if inserted successfuly, otherwise <B>false</B>
     * is returned (the entry may already be there).
     */
    public boolean insertEntry(byte[] oid, String key) {

        Assert.check( entries!=null );

        readWriteLock.writeLock();
        this.entries.insertEntry(oid,key);
        readWriteLock.writeUnlock();
        return true;

    }

    /**
     * Removes the entry with the specified <I>oid</I>
     * from the internal database.
     *
     * @param oid The object Id field of the pair to be removed.
     *
     * @return <B>true</B> if removed successfuly, otherwise <B>false</B>
     * is returned (the entry may not be present).
     */
    public boolean removeEntry(byte[] oid) {

        Assert.check( entries!=null );

        readWriteLock.writeLock();
        this.entries.removeByObjectId(oid);
        readWriteLock.writeUnlock();
        return true;

    }

    /**
     * Removes the entry with the specified <I>key</I>
     * from the internal database.
     *
     * @param key The object key field of the pair to be removed.
     *
     * @return <B>true</B> if removed successfuly, otherwise <B>false</B>
     * is returned (the entry may not be present).
     */
    public boolean removeEntry(String key) {

        Assert.check( entries!=null );

        readWriteLock.writeLock();
        this.entries.removeByKey(key);
        readWriteLock.writeUnlock();
        return true;

    }

    /**
     * This tests whether a entry with the specified <I>key</I> is present
     * in the internal database.
     *
     * @param key The object key of the requested entry.
     *
     * @return <B>true</B> if an entry with the specified <I>key</I> field is
     * present in the internal database. Otherwise return <B>false</B>.
     */
    public boolean contains(String key) {

        Assert.check( entries!=null );

        readWriteLock.readLock();
        boolean result = this.entries.containsKey(key);
        readWriteLock.readUnlock();

        return result;

    }

    /**
     * This tests whether a entry with the specified <I>oid</I> is present
     * in the internal database.
     *
     * @param oid The object Id of the requested entry.
     *
     * @return <B>true</B> if an entry with the specified <I>oid</I> field is
     * present in the internal database. Otherwise return <B>false</B>.
     */
    public boolean contains(byte[] oid) {

        Assert.check( entries!=null );

        readWriteLock.readLock();
        boolean result = this.entries.containsObjectId(oid);
        readWriteLock.readUnlock();

        return result;

    }

    /**
     * Returns the number of entries in the internal database.
     *
     * @return The number of entries.
     */
    public int entries() {

        Assert.check( entries!=null );

        readWriteLock.readLock();
        int result = this.entries.size();
        readWriteLock.readUnlock();

        return result;

    }

    /**
     * This dumps content of the internal database to the specified
     * output stream.
     *
     * @param out A reference to an output printstream.
     */
    public void dump(java.io.PrintStream out) {

        Assert.check( entries!=null );

        readWriteLock.readLock();
        out.print("EntityCleanupNofifier  --- Size = " + entries() + "\n");
        out.print(this.entries.dump());
        readWriteLock.readUnlock();

    }

    /**
     * This operation initialises internal database using a property as
     * defined by the OMG CosPropertyService.
     *
     * @param propertySet  A reference to the PropertySet containing
     * the property.
     * @param propertyName Name of the property within the property set above.
     *
     * @return <B>true</B> if <I>init</I> was successful,
     * <B>false</B> otherwise.
     */
    public boolean init(PropertySet propertySet, String propertyName) {

        boolean result = false;

        if ( propertySet!=null ) {
            Assert.check(propertyName.length()!=0);

            // The follows steps are performed:
            // 1. Get value of property named "propetyName"
            // 2. For each value, insertEntry to internal entries database
            // 3. If PropertyNotFound exception is raised,
            //      it means it's first usage.
            //    Do not create a property at this time but do it
            //    in finish() before persisting state.
            try {
                this.propertySet = propertySet;
                this.propertyName = propertyName;

                org.omg.CORBA.Any any
                    = this.propertySet.get_property_value(propertyName);

                // Data is stored as sequence {oid;key}
                EntityDatum[] eData = EntityDataHelper.extract(any);

                for (int i = 0; i < eData.length; ++i) {
                    insertEntry(eData[i].object_id,eData[i].object_key);
                }
                result = true;
            } catch(PropertyNotFound pnf) {
                // Property not found, create at later!
                result = true;
            } catch(InvalidPropertyName ipn) {
                result = true;
            } catch(Exception e) {
                result = false;
            }
        }

        return result;

    }

    /**
     * This operation instructs the EntityCleanupNotifier that
     * the LifeCycle framework is done with it and can then persist its state
     * for next sessions.
     *
     * @param orb ORB reference used to create a Any object
     *
     * @return <B>true</B> if done successfuly, <B>false</B> otherwise.
     */
    public boolean finish(org.omg.CORBA.ORB orb) {

        boolean result = false;
        if (this.propertySet!=null) {
            try {
                // Create a CORBA Any containing current state
                readWriteLock.readLock();
                EntityDatum[] ed = this.entries.toEntityData();

                org.omg.CORBA.Any any = orb.create_any();
                EntityDataHelper.insert(any, ed);

                // This will either define a new property or set a new value
                // for the property.
                this.propertySet.define_property(this.propertyName, any);
                // State is persisted!
                result = true;
            } catch (org.omg.CORBA.UserException ue) {
                // Possible exceptions are:
                //        InvalidPropertyName,
                //        ConflictingProperty,
                //        UnsupportedTypeCode,
                //        UnsupportedProperty,
                //        ReadOnlyProperty

                // do nothing for this version
            } finally {
                readWriteLock.readUnlock();
            }

        }

        return result;

    }

    /**
     * This returns the <I>key</I> value of the entry with the specified
     * object Id.
     *
     * @param oid Object Id corresponding to the string returned.
     *
     * @return The object key value of the entry.
     */
    protected String objectIdToKey(byte[] oid) {

        Assert.check( entries!=null );

        readWriteLock.readLock();
        String key = this.entries.getKey(oid);
        readWriteLock.readUnlock();

        return key;

    }

    /**
     * This returns the Object Id value of the entry with the specified
     * object key.
     *
     * @param key Object key corresponding to the Object Id returned.
     *
     * @return The Object Id value of the entry.
     */
    protected byte[] keyToObjectId(String key) {

        Assert.check( entries!=null );

        readWriteLock.readLock();
        byte[] oid = this.entries.getObjectId(key);
        readWriteLock.readUnlock();

        return oid;

    }

}

File: LifeCycle/lifecycle/c++/src/EntityCleanupNotifier.cpp
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES 
 * and SELEX-SI. 
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003. 
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under 
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your 
 * option) any later version. 
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public 
 * License for more details. 
 * 
 * You should have received a copy of the GNU Library General 
 * Public License along with CARDAMOM; see the file COPYING. If not, write to 
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


#include "LifeCycle/lifecycle/EntityCleanupNotifier.hpp"
#include <map>
#include <algorithm>
#include "Foundation/common/Assert.hpp"
#include "Foundation/osthreads/ReaderLockGuard.hpp"
#include "Foundation/osthreads/WriterLockGuard.hpp"
#include "lifecycle/CdmwEntityData.stub.hpp"

namespace
{

    /**
     * Class responsible for dumping content of an entry to an output stream
     */

    class EntryDumper
    {

        public:
            EntryDumper(std::ostream & o)
                    : m_os(o)
            {}

            void operator()(
                const std::pair<const std::string , const std::string > & p
            ) const
            {
                m_os << "<ObjectId = " << p.first << ", Key = " << p.second << ">\n";
            }

        private:
            std::ostream & m_os;
    };

    /**
     * Class responsible for appending to a CdmwLifeCycle::EntityData sequence
     * a pait of <oid,key>.
     */

    class EntityDataFiller
    {

        public:
            EntityDataFiller(CdmwLifeCycle::EntityStrategy::EntityData & ed)
                    : m_ed(ed)
            {}

            void operator()(const std::pair<const std::string, const std::string > & p)
            const
            {
                CORBA::ULong ed_lg = m_ed.length();
                m_ed.length(ed_lg + 1);
                PortableServer::ObjectId_var oid
                = PortableServer::string_to_ObjectId(p.first.c_str());
                CORBA::ULong lg = oid->length();
                m_ed[ed_lg].object_id.length(lg);

                for (CORBA::ULong i = 0; i < lg;++i)
                    m_ed[ed_lg].object_id[i] = oid[i];

                m_ed[ed_lg].object_key = CORBA::string_dup(p.second.c_str());
            }

        private:
            CdmwLifeCycle::EntityStrategy::EntityData & m_ed;
    };
}

; // End anonymous namespace

namespace Cdmw
{

    namespace LifeCycle
    {

        /**
        *Implementation notes: [if no pertinent write none ]
        *<p>
        *Portability issues: [if no pertinent write none ]
        *<p>
        */


        class EntityCleanupNotifier::EntriesMap :
                    public std::map<std::string, std::string> // map<oid,key>
        {

            public:
                typedef value_type element_type;

                PortableServer::ObjectId*
                get_ObjectId(const element_type & elem) const
                {
                    PortableServer::ObjectId_var oid
                    = PortableServer::string_to_ObjectId(elem.first.c_str());
                    return oid._retn();
                }

                std::string
                get_key(const element_type & elem) const
                {
                    return elem.second;
                }

                const_iterator
                find_by_key(const std::string & key) const throw()
                {
                    return std::find_if(begin(), end(), KeyFinder(key));
                }

                iterator
                find_by_key(const std::string & key) throw()
                {
                    return std::find_if(begin(), end(), KeyFinder(key));
                }

                const_iterator
                find_by_ObjectId(const PortableServer::ObjectId & oid) const throw()
                {
                    const_iterator where;

                    try
                    {
                        CORBA::String_var s = PortableServer::ObjectId_to_string(oid);
                        std::string s_oid(s.in());

                        where = find(s_oid);
                    }
                    catch (...)
                    {
                        where = end();
                    }

                    return where;
                }

                iterator
                find_by_ObjectId(const PortableServer::ObjectId & oid) throw()
                {
                    iterator where;

                    try
                    {
                        CORBA::String_var s = PortableServer::ObjectId_to_string(oid);
                        std::string s_oid(s.in());

                        where = find(s_oid);
                    }
                    catch (...)
                    {
                        where = end();
                    }

                    return where;
                }

                bool
                insert_entry(const PortableServer::ObjectId & oid, const std::string & key) throw()
                {
                    bool done = false;

                    try
                    {
                        CORBA::String_var s = PortableServer::ObjectId_to_string(oid);
                        std::string s_oid(s.in());

                        iterator where = find(s_oid);

                        if (where == end())
                        {
                            insert(std::make_pair<std::string, std::string>(s_oid, key));
                            done = true;
                        }
                    }
                    catch (...)
                    {}

                    return done;
                }

                bool
                remove_by_ObjectId(const PortableServer::ObjectId & oid) throw()
                {
                    bool done = false;

                    try
                    {
                        CORBA::String_var s = PortableServer::ObjectId_to_string(oid);
                        std::string s_oid(s.in());

                        iterator where = find(s_oid);

                        if (where != end())
                        {
                            erase(s_oid);
                            done = true;
                        }
                    }
                    catch (...)
                    {}

                    return done;
                }

                bool
                remove_by_key(const std::string & key) throw()
                {
                    bool done = false;

                    try
                    {

                        iterator where = find_by_key(key);

                        if (where != end())
                        {
                            erase(where->first);
                            done = true;
                        }
                    }
                    catch (...)
                    {}

                    return done;
                }

            private:
                // Function object for the find_if algorithm to search for
                // entries by object_key instead of ObjectId.

                class KeyFinder
                {

                    public:
                        KeyFinder(const std::string & key)
                                : m_key(key)
                        {}

                        bool
                        operator()(const std::pair<const std::string , const std::string > & p) const
                        {
                            return p.second == m_key;
                        }

                    private:
                        const std::string & m_key;
                };
        };



        EntityCleanupNotifier::EntityCleanupNotifier(CdmwLifeCycle::ObjectRegistration_ptr domain)
        throw(CORBA::SystemException,
              Cdmw::OutOfMemoryException)
                : m_domain(CdmwLifeCycle::ObjectRegistration::_duplicate(domain)),
                m_entries(0),
                m_property_set(CosPropertyService::PropertySet::_nil()),
                m_property_name("")
        {
            try
            {
                std::auto_ptr<EntriesMap> entries(new EntriesMap());
                m_entries = entries;
            }
            catch (const std::bad_alloc &)
            {
                CDMW_THROW(Cdmw::OutOfMemoryException);
            }
        }

        EntityCleanupNotifier::~EntityCleanupNotifier()
        throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);
        }

        void EntityCleanupNotifier::unregister_object(const PortableServer::ObjectId & oid)
        throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);

            CDMW_WRITER_LOCK_GUARD(m_readers_writer_lock);

            EntriesMap::iterator where = m_entries->find_by_ObjectId(oid);

            if (where != m_entries->end())
            {
                try
                {
                    std::string key = m_entries->get_key(*where);
                    m_domain->unregister_object(key.c_str());
                    m_entries->erase(where);
                }
                catch (const CosNaming::NamingContext::NotFound & )
                {
                    // This should not happen (TODO): ADD Trace logs
                }
                catch (const CosNaming::NamingContext::InvalidName & )
                {
                    // This should not happen either (TODO): ADD Trace logs
                }
                catch (const CORBA::SystemException & )
                {
                    // What to do!!! (TODO): ADD Trace logs
                }
            }
            else
            {
                // Ignore
            }

        }


        // Inserts a (oid,key) pair. Returns false if entry is already there.
        bool
        EntityCleanupNotifier::insert_entry(const PortableServer::ObjectId & oid,
                                            const std::string & key)
        throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);

            CDMW_WRITER_LOCK_GUARD(m_readers_writer_lock);
            return m_entries->insert_entry(oid, key);
        }

        // Removes the (oid,key) pair.
        bool
        EntityCleanupNotifier::remove_entry(const PortableServer::ObjectId & oid)
        throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);

            CDMW_WRITER_LOCK_GUARD(m_readers_writer_lock);
            return m_entries->remove_by_ObjectId(oid);
        }

        bool
        EntityCleanupNotifier::remove_entry(const std::string & key)
        throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);

            CDMW_WRITER_LOCK_GUARD(m_readers_writer_lock);
            return m_entries->remove_by_key(key);
        }

        bool
        EntityCleanupNotifier::contains(const std::string & key) const
        throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);

            CDMW_READER_LOCK_GUARD(m_readers_writer_lock);
            EntriesMap::const_iterator where = m_entries->find_by_key(key);
            return where != m_entries->end();
        }

        bool
        EntityCleanupNotifier::contains(const PortableServer::ObjectId & oid) const
        throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);

            CDMW_READER_LOCK_GUARD(m_readers_writer_lock);
            EntriesMap::const_iterator where = m_entries->find_by_ObjectId(oid);
            return where != m_entries->end();
        }

        // Returns Number of entries
        size_t
        EntityCleanupNotifier::entries() const throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);

            CDMW_READER_LOCK_GUARD(m_readers_writer_lock);
            return m_entries->size();
        }

        void
        EntityCleanupNotifier::dump(std::ostream & os) const throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);

            CDMW_READER_LOCK_GUARD(m_readers_writer_lock);
            os << "EntityCleanupNofifier  --- Size = " << m_entries->size() << "\n";
            std::for_each(m_entries->begin(), m_entries->end(), EntryDumper(os));
        }


        // These should only be called when associated entry does exist!
        std::string
        EntityCleanupNotifier::ObjectId_to_key(const PortableServer::ObjectId& oid) const
        throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);

            CDMW_READER_LOCK_GUARD(m_readers_writer_lock);
            EntriesMap::const_iterator where = m_entries->find_by_ObjectId(oid);
            CDMW_ASSERT(where != m_entries->end());

            return m_entries->get_key(*where);
        }

        PortableServer::ObjectId*
        EntityCleanupNotifier::key_to_ObjectId(const std::string & key) const
        throw()
        {
            CDMW_ASSERT(m_entries.get() != NULL);

            CDMW_READER_LOCK_GUARD(m_readers_writer_lock);
            EntriesMap::const_iterator where = m_entries->find_by_key(key);
            CDMW_ASSERT(where != m_entries->end());

            return m_entries->get_ObjectId(*where);
        }

        bool
        EntityCleanupNotifier::init(CosPropertyService::PropertySet_ptr property_set,
                                    const std::string & property_name)
        throw()
        {
            bool result = false;

            if (!CORBA::is_nil(property_set))
            {
                CDMW_ASSERT(!property_name.empty());

                // The follows steps are performed:
                // 1. Get value of property named "propety_name"
                // 2. For each value, insert_entry to internal entries database
                // 3. If PropertyNotFound exception is raised, it means it's first usage.
                //    Do not create a property at this time but do it in finish() before
                //    persisting state.

                try
                {
                    m_property_set = CosPropertyService::PropertySet::_duplicate(property_set);
                    m_property_name = property_name;

                    CORBA::Any_var any
                    = property_set->get_property_value(property_name.c_str());

                    // Data is stored as sequence<struct{oid;key}>
                    const CdmwLifeCycle::EntityStrategy::EntityData* pData;
                    any.in() >>= pData;

                    for (CORBA::ULong i = 0; i < pData->length(); ++i)
                    {
                        const CORBA::Octet * buff = (*pData)[i].object_id.get_buffer();
                        CORBA::ULong lg = (*pData)[i].object_id.length();

                        // Avoid copy
                        PortableServer::ObjectId oid(lg,
                                                     lg,
                                                     const_cast<CORBA::Octet*>(buff),
                                                     false);
                        insert_entry(oid, (*pData)[i].object_key.in());
                    }

                    result = true;
                }
                catch (const CosPropertyService::PropertyNotFound&)
                {
                    // Property not found, create at later!
                    result = true;
                }
                catch (...)
                {
                    // Possible exceptions are std::bad_alloc and CORBA::SystemException
                    result = false;
                }
            }

            return result;
        }


        bool
        EntityCleanupNotifier::finish()
        throw()
        {
            bool result = false;

            if (!CORBA::is_nil(m_property_set.in()))
            {
                try
                {
                    // Create a CORBA::Any containing current state
                    CDMW_READER_LOCK_GUARD(m_readers_writer_lock);
                    CORBA::ULong lg = m_entries->size();
                    CdmwLifeCycle::EntityStrategy::EntityData * ed =
                        new CdmwLifeCycle::EntityStrategy::EntityData(lg);
                    ed->length(0);
                    std::for_each(m_entries->begin(), m_entries->end(), EntityDataFiller(*ed));

                    CORBA::Any any;
                    any <<= ed; // consuming insertion

                    // This will either define a new property or set a new value
                    // for the property.
                    m_property_set->define_property(m_property_name.c_str(), any);
                    // State is persisted!
                    result = true;
                }
                catch (const CORBA::Exception &)
                {
                    // Possible exceptions are:
                    //    . CosPropertyService::ReadOnlyProperty
                    //    . CosPropertyService::ConflictingProperty
                    //    . CosPropertyService::UnsupportedProperty
                    //    . CosPropertyService::UnsupportedTypeCode
                    //    . CosPropertyService::InvalidPropertyName
                    //    . CORBA::SystemException

                    // do nothing for this version
                }

            }

            return result;
        }


    }

    ; // End namespace LifeCycle
}

; // End namespace Cdmw



